File: src/dotnet/HQ.Abstractions/Times/GetDashboardTimeV1.cs
using System.Diagnostics;

using HQ.Abstractions.Enumerations;

using static HQ.Abstractions.Times.GetTimesV1;

namespace HQ.Abstractions.Times;

public class GetDashboardTimeV1
{
    public class Request
    {
        public Guid StaffId { get; set; }
        public DateOnly Date { get; set; }
        public Period Period { get; set; }
        public SortColumn SortBy { get; set; } = SortColumn.Date;
        public SortDirection SortDirection { get; set; } = SortDirection.Desc;

        public string? Search { get; set; }
        public TimeStatus? Status { get; set; }
    }

    public class Response
    {
        public decimal TotalHours { get; set; }
        public decimal BillableHours { get; set; }
        public DateOnly StartDate { get; set; }
        public DateOnly EndDate { get; set; }
        public DateOnly PreviousDate { get; set; }
        public DateOnly NextDate { get; set; }
        public List<TimeForDate> Dates { get; set; } = new();
        public decimal HoursThisWeek { get; set; }
        public decimal HoursThisMonth { get; set; }
        public decimal HoursLastWeek { get; set; }
        public string? StaffName { get; set; }
        public decimal Vacation { get; set; }
        public int RejectedCount { get; set; }
        public bool CanSubmit { get; set; }

        public DateOnly? TimeEntryCutoffDate { get; set; }
    }

    public class TimeForDate
    {
        public bool CanCreateTime { get; set; }
        public DateOnly Date { get; set; }
        public DateOnly StartDate { get; set; }

        public DateOnly EndDate { get; set; }

        public List<TimeEntry> Times { get; set; } = new List<TimeEntry>();
        public decimal TotalHours { get; set; }
    }

    public class TimeEntry
    {
        public Guid Id { get; set; }
        public DateTime CreatedAt { get; set; }
        public DateOnly Date { get; set; }
        public decimal Hours { get; set; }
        public decimal MaximumTimeEntryHours { get; set; }
        public string? Notes { get; set; }
        public string? Task { get; set; }
        public Guid? ChargeCodeId { get; set; }
        public Guid? ClientId { get; set; }
        public Guid? ProjectId { get; set; }
        public Guid? ActivityId { get; set; }
        public string? ActivityName { get; set; }
        public string? ProjectName { get; set; }
        public string? ClientName { get; set; }
        public TimeStatus TimeStatus { get; set; }
        public string ChargeCode { get; set; } = null!;
        public string? RejectionNotes { get; set; }
    }
}

File: src/angular/hq/src/app/models/staff-dashboard/get-dashboard-time-v1.ts
import { Period } from '../../enums/period';
import { TimeStatus } from '../../enums/time-status';

export interface GetDashboardTimeV1Request {
  staffId: string;
  period: Period;
  date?: string;
  search?: string | null;
  status?: TimeStatus | null;
}

export interface GetDashboardTimeV1Response {
  totalHours: number;
  billableHours: number;
  hoursThisWeek: number;
  hoursThisMonth: number;
  hoursLastWeek: number;
  maximumTimeEntryHours: number;
  staffName: string;
  vacation: number;
  startDate: string;
  endDate: string;
  previousDate: string;
  nextDate: string;
  dates: GetDashboardTimeV1TimeForDate[];
  rejectedCount: number;
  canSubmit: boolean;
  timeEntryCutoffDate: string;
}

export interface GetDashboardTimeV1TimeForDate {
  date: string;
  times: GetDashboardTimeV1TimeForDateTimes[];
  totalHours: number;
  canCreateTime: boolean;
}

export interface GetDashboardTimeV1TimeForDateTimes {
  id: string;
  date: string;
  hours: number;
  notes: string | null;
  task: string | null;
  chargeCodeId: string | null;
  maximumTimeEntryHours: number;
  chargeCode: string;
  clientId: string | null;
  projectId: string | null;
  activityName: string | null;
  clientName: string | null;
  projectName: string | null;
  activityId: string | null;
  timeStatus: TimeStatus | null;
  rejectionNotes: string | null;
}
